### PR TITLE
Fix autograd when `inputs` contains tensors without materialized grad_fn 

### LINKED ISSUE #39784: Autograd always backpropagates to arithmetic operations when there is an unused input
## 🐛 Bug

If:
- a simple arithmetic operation (add, multiply) is used in the computation graph
- AND `autograd.grad` is called wrt an unused input

then:
- autograd will include this arithmetic operation (and everything leading back to it) in its backward pass.

Remarkably, this occurs even if `allow_unused=False`. The backpropagation is performed first, and the error about unused inputs is thrown afterwards!

I've not tested every operation (subtraction, division, etc.), but I can confirm that this bug doesn't occur with certain other operations (clone, relu, view).

## To Reproduce

```python
import torch

class Print(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x, op, final, unused):
        ctx.op = op
        ctx.final = final
        ctx.unused = unused
        return x

    @staticmethod
    def backward(ctx, x):
        print('op={}, final={}, unused={} has a bug'.format(ctx.op, ctx.final, ctx.unused))
        return x, None, None, None

def f(op, final, unused):
    x = torch.rand(1, requires_grad=True)
    if op == 'mul':
        x = 2 * x
    elif op == 'add':
        x = 2 + x
    elif op == 'relu':
        x = x.relu()
    elif op == 'clone':
        x = x.clone()
    elif op == 'view':
        x = x.view(-1)
    x = Print.apply(x, op, final, unused)

    y = torch.rand(1, requires_grad=True)

    if final:
        x = y + x

    try:
        torch.autograd.grad((x,), (y,), allow_unused=unused)
    except RuntimeError:
        pass


f('mul', False, True)
f('add', False, True)
f('relu', False, True)
f('clone', False, True)
f('view', False, True)

f('mul', True, True)
f('add', True, True)
f('relu', True, True)
f('clone', True, True)
f('view', True, True)

f('mul', True, False)
f('add', True, False)
f('relu', True, False)
f('clone', True, False)
f('view', True, False)

f('mul', False, False)
f('add', False, False)
f('relu', False, False)
f('clone', False, False)
f('view', False, False)
```

Prints:
```
op=mul, final=False, unused=True has a bug
op=add, final=False, unused=True has a bug
op=mul, final=False, unused=False has a bug
op=add, final=False, unused=False has a bug
```

## Expected behavior

Nothing to be printed at all!

## Environment

 - PyTorch Version (e.g., 1.0): 1.5
 - OS (e.g., Linux): Ubuntu 18.04 LTS
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Python version: 3.8.2

## Additional context

This came up in a context where I'm handling the backward computation quite carefully - and this bug is currently producing several-order-of-magnitude slowdowns because the autograd engine is backprop'ing through very expensive parts of the graph that it really shouldn't be touching.

cc @ezyang @SsnL @albanD @zou3519 @gqchen

### PR DESCRIPTION
**Pull request by soulitzer**
Fixes #39784
At the time the issue was filed, there was only issue (1) below. 

There are actually now two issues here:
1. We always set all inputs passed in through `inputs` arg as `needed = True` in exec_info. So if we pass in an input that has a grad_fn that is not materialized, we create an entry of exec_info with nullptr as key with `needed = True`. Coincidentally, when we perform simple arithmetic operations, such as "2 * x", one of the next edges of mul is an invalid edge, meaning that its grad_fn is also nullptr. This causes the discovery algorithm to set all grad_fns that have a path to this invalid_edge as `needed = True`.
2. Before the commit that enabled the engine skipped the dummy node, we knew that root node is always needed, i.e., we hardcode `exec_info[&graph_root]=true`. The issue was that this logic wasn't updated after the code was updated to skip the graph root.

To address (1), instead of passing in an invalid edge if an input in `inputs` has no grad_fn, we create a dummy grad_fn. This is done in both python and cpp entry points. The alternative is to add logic for both backward() and grad() cases to check whether the grad_fn is nullptr and set needed=false in that case (the .grad() case would be slightly more complicated than the .backward() case here).

For (2), we perform one final iteration of the discovery algorithm so that we really know whether we need to execute the graph root.

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -600,6 +600,35 @@ def test_grad_unreachable(self):
...
+        x = MyFunc.apply(torch.randn(1, requires_grad=True) * 2)
+        y = torch.randn(1, requires_grad=True)
+        z = torch.randn(1, requires_grad=True)
+        (gY, gZ) = torch.autograd.grad(x + z, (y, z), allow_unused=True)
+        self.assertIsNone(gY)
+        self.assertIsNotNone(gZ)
+
+        x = MyFunc.apply(torch.randn(1, requires_grad=True) * 2)
+        y = torch.randn(1, requires_grad=True)
+        torch.autograd.backward(x, inputs=(y, ))  # allow_unused is implicitly True!
```
Comment (albanD) [inline test/test_autograd.py:629]: You mean False?

File: torch/csrc/autograd/autograd.cpp [hunk 2]
```diff
@@ -4,6 +4,7 @@
 #include <torch/csrc/autograd/edge.h>
 #include <torch/csrc/autograd/engine.h>
 #include <torch/csrc/autograd/function.h>
+#include <torch/csrc/autograd/functions/basic_ops.h>
```
Comment (albanD) [inline torch/csrc/autograd/autograd.cpp:7]: Why the extra include?

File: torch/csrc/autograd/python_engine.cpp [hunk 3]
```diff
@@ -237,7 +238,13 @@ PyObject *THPEngine_run_backward(PyObject *self, PyObject *args, PyObject *kwarg
       THPUtils_assert(input_var->cdata.requires_grad(),
           "One of the differentiated Tensors does not require grad");
       if (!grad_fn) {
-        output_edges.emplace_back();
+        // Since input has no grad_accumulator, its guaranteed to be not reachable.
```
Comment (albanD) [inline torch/csrc/autograd/python_engine.cpp:241]: nit: unreachable

File: torch/csrc/autograd/python_engine.cpp [hunk 4]
```diff
@@ -237,7 +238,13 @@ PyObject *THPEngine_run_backward(PyObject *self, PyObject *args, PyObject *kwarg
...
+        // We create a dummy edge with no next edges and a non-nullptr grad_fn so
+        // nodes (e.g., mul when an operand is scalar) that have invalid next_edges
```
Comment (albanD) [inline torch/csrc/autograd/python_engine.cpp:243]: I'm not sure if I would say invalid here. Maybe empty?
Because even in c++ if you add two Tensors, one that requires grad and one that does not, one of the edge will point nowhere for the Tensor that does not require gradients.

File: torch/csrc/autograd/python_engine.cpp [hunk 5]
```diff
@@ -237,7 +238,13 @@ PyObject *THPEngine_run_backward(PyObject *self, PyObject *args, PyObject *kwarg
...
+        // don't get erroneously assigned `needed = True` in exec_info. Note that this
+        // doesn't happen in the cpp api because scalars are only wrapped into tensors
```
Comment (albanD) [inline torch/csrc/autograd/python_engine.cpp:245]: Not sure this is true when you mix tensors that do and don't require gradients, see above.

File: torch/csrc/autograd/python_engine.cpp [hunk 6]
```diff
@@ -237,7 +238,13 @@ PyObject *THPEngine_run_backward(PyObject *self, PyObject *args, PyObject *kwarg
...
+        // in python
+        output_edges.emplace_back(std::make_shared<Identity>(), 0);
```
Comment (albanD) [inline torch/csrc/autograd/python_engine.cpp:247]: nit: you could create a basic `Node` here no? Or does the fact that it is an `Identity` important?

File: test/test_autograd.py (hunk 1 above)
Comment (soulitzer) [reply test/test_autograd.py:629]: It is False by default for .grad() but seems to be True for .backward() since its does not raise an error here

File: torch/csrc/autograd/python_engine.cpp (hunk 5 above)
Comment (soulitzer) [reply torch/csrc/autograd/python_engine.cpp:245]: Ahh, forgot about this case. Good catch!

File: torch/csrc/autograd/autograd.cpp (hunk 2 above)
Comment (soulitzer) [reply torch/csrc/autograd/autograd.cpp:7]: Oops forgot to remove this, but turns out we might need it now?

File: test/test_autograd.py (hunk 1 above)
Comment (albanD) [reply test/test_autograd.py:629]: Ho right sorry. I though this was a call to autograd.grad because you are mentioning this flag.
It does not exist for backward, so yes it is as if it was always True.

File: torch/csrc/autograd/autograd.cpp (hunk 2 above)
Comment (albanD) [reply torch/csrc/autograd/autograd.cpp:7]: You don't add anything to this file. So I don't think you need it here no? Maybe new includes in python_engine yes :)

File: torch/csrc/autograd/python_engine.cpp (hunk 6 above)
Comment (soulitzer) [reply torch/csrc/autograd/python_engine.cpp:247]: It doesn't have to be `Identity` here but I don't think we can directly initialize the abstract Node class here

File: torch/csrc/autograd/python_engine.cpp (hunk 4 above)
Comment (soulitzer) [reply torch/csrc/autograd/python_engine.cpp:243]: Yeah, invalid isn't very clear here

File: torch/csrc/autograd/autograd.cpp (hunk 2 above)
Comment (soulitzer) [reply torch/csrc/autograd/autograd.cpp:7]: I actually did the fix for cpp run_backward entry point originally as well, but when I ported the same tests over it looked like they were passing even without the fix. So I removed the fix, but left this import on accident :P
From your other comments though it looks like we need the fix for the cpp api after all since there is also the requires_grad/non-requires grad case that I didn't consider.

File: torch/csrc/autograd/python_engine.cpp (hunk 4 above)
Comment (soulitzer) [reply torch/csrc/autograd/python_engine.cpp]: Changing it to say: "edge pointing to nullptr"

File: test/test_autograd.py (hunk 1 above)
Comment (soulitzer) [reply test/test_autograd.py:629]: Would it also make sense to have something like "allow_unused" for backward for symmetry? I guess its going to be tough making the default False at this point bc it would be bc breaking, but seems like by default unused inputs should raise an inputs for backward as well.

Comment (albanD) [reply test/test_autograd.py:629]: I am not sure we want to add that flag.
Also it might be tricky to implement because we don't capture anything in this case and so it will be hard to know if the Nodes were used or not.

File: torch/csrc/autograd/python_engine.cpp (hunk 6 above)
Comment (albanD) [reply torch/csrc/autograd/python_engine.cpp:245]: Ho it is abstract. Ok!

File: torch/csrc/autograd/autograd.cpp [hunk 7]
```diff
@@ -103,7 +104,8 @@ variable_list run_backward(
           input.requires_grad(),
           "One of the differentiated Tensors does not require grad");
       if (!grad_fn) {
-        output_edges.emplace_back();
+        // See the python_engine's run_backward for an explanation
```
Comment (albanD) [inline torch/csrc/autograd/autograd.cpp:107]: nit: when we want to do these references, we usually name these like that:
https://github.com/pytorch/pytorch/blob/1c8d11c9e2ff64d92fb322c573d287588c26a718/torch/csrc/autograd/variable.h#L330

And reference them with that name (so that grepping for the note is easy)
https://github.com/pytorch/pytorch/blob/1c8d11c9e2ff64d92fb322c573d287588c26a718/torch/csrc/autograd/variable.h#L98

**Review by albanD — APPROVED** [review]
LGTM!

File: torch/csrc/autograd/autograd.cpp (hunk 7 above)
Comment (soulitzer) [reply torch/csrc/autograd/autograd.cpp:107]: Cool!
